File: backend/schemas/practice.py
from pydantic import BaseModel
from typing import Optional

class StartPracticeRequest(BaseModel):
    questionCount: int
    questionBankId: str | None = None


class StartPracticeResponse(BaseModel):
    success: bool
    practiceId: str
    questionCount: int
    questionBank: str


class SubmitPracticeAnswerRequest(BaseModel):
    questionId: str
    selectedOption: str


class SubmitPracticeAnswerResponse(BaseModel):
    correct: bool
    correctOption: str
    explanation: Optional[str] = None


class FinishPracticeResponse(BaseModel):
    correct: int
    wrong: int
    skipped: int
    accuracy: float

File: frontend/src/types/practice.ts
export interface PracticeQuestion {
  id: string;
  number: number;
  subject?: string;

  question: string;

  options: Record<string, string>;

  correctAnswer: string;

  explanation?: string;
}

export interface PracticeAnswer {
  selectedOption: string;
}

export interface PracticeSession {
  practiceId: string;

  questionBank: string;

  questionCount: number;

  questions: PracticeQuestion[];

  answers: Record<string, PracticeAnswer>;
}

export interface StartPracticeResponse {
  success: boolean;

  practiceId: string;

  questionCount: number;

  questionBank: string;
}

export interface SubmitPracticeAnswerResponse {
  correct: boolean;

  correctAnswer: string;

  explanation?: string;
}

export interface FinishPracticeResponse {
  correct: number;

  wrong: number;

  skipped: number;

  accuracy: number;
}